sherpa/cpp_api/brpc/server.cc

```diff
@@ -0,0 +1,380 @@
+      {
+        if (vad_state_queue_[stream_id] == endpoint_reset) {
+          stream_queue_[stream_id] = online_asr_.CreateStream();
+          vad_state_queue_[stream_id] = endpoint_inactive;
+        }
+        const int num_samples = len / sizeof(int16_t);
+        char * pcm_data = const_cast<char *>(data);
+        auto wav_stream_tensor = torch::from_blob(
+            reinterpret_cast<int16_t *>(pcm_data),
+            {num_samples},
+            torch::kInt16).to(torch::kFloat) / 32768;
+
+        stream_queue_[stream_id]->AcceptWaveform(SAMPLE_RATE,
+            wav_stream_tensor);
+        if (end || vad_state_queue_[stream_id] == input_end) {
+          stream_queue_[stream_id]->InputFinished();
+          vad_state_queue_[stream_id] = endpoint_reset;
+        }
+      }
+      mlock.unlock();
+      batch_cond_.notify_one();
+      if (vad_state_queue_[stream_id] != endpoint_reset) {
+        if (segment_stream(stream_id)) {
+          vad_state_queue_[stream_id] = endpoint_active;
+        }
+      }
+    }
+  }
+
+  const char * fetch_stream(const int64_t stream_id) {
+    std::shared_lock mlock(mutex_);
+    if (stream_queue_.find(stream_id) != stream_queue_.end()) {
+      {
+        decoding_cond_.wait_for(mlock, wait_ms_ * 1ms,
+            [this] { return !decoding_; });
+        std::string result =
+          online_asr_.GetResult(stream_queue_[stream_id].get());
+        result_queue_[stream_id].resize(MAX_ASR_RESULT);
+        strncpy(result_queue_[stream_id].data(),
+            result.c_str(),
+            std::min(static_cast<int>(result.size()), MAX_ASR_RESULT));
+      }
+      mlock.unlock();
+      if (segment_stream(stream_id)) {
+        vad_state_queue_[stream_id] = endpoint_active;
+      }
+      return result_queue_[stream_id].data();
+    }
+    return nullptr;
+  }
+
+  int32_t segment_stream(const int64_t stream_id) {
+    std::shared_lock mlock(mutex_);
+    if (stream_queue_.find(stream_id) != stream_queue_.end()) {
+      return stream_queue_[stream_id]->IsEndpoint();
+    }
+    return false;
+  }
+
+  void batch_decode() {
+    while (alive_) {
+      {
+        std::unique_lock mlock(mutex_);
+        batch_cond_.wait_for(mlock, wait_ms_ * 1ms,
+            [this] { return batch_size_ <= stream_queue_.size(); });
+
+        std::vector<sherpa::OnlineStream *> ready_streams;
+        for (const auto & stream : stream_queue_) {
+          if (online_asr_.IsReady(stream.second.get())) {
+            ready_streams.push_back(stream.second.get());
+          }
+        }
+
+        if (ready_streams.empty()) { continue; }
+        decoding_ = true;
+        online_asr_.DecodeStreams(ready_streams.data(), ready_streams.size());
+        decoding_ = false;
+      }
+      decoding_cond_.notify_all();
+    }
+    std::this_thread::sleep_for(std::chrono::milliseconds(100));
+    exit_ = true;
+  }
+
+  void stop_decode() {
+    std::unique_lock mlock(mutex_);
```

Comment:
```suggestion
    std::lock_guard mlock(mutex_);
```

If you need to transfer the ownership, then you can use `unique_xxx`, e.g., `std::unique_ptr`, `std::unique_lock`.